packages/contracts-bedrock/contracts/echidna/FuzzResourceMetering.sol

```diff
@@ -0,0 +1,133 @@
+pragma solidity 0.8.15;
+
+import { ResourceMetering } from "../L1/ResourceMetering.sol";
+import { Arithmetic } from "../libraries/Arithmetic.sol";
+import { StdUtils } from "forge-std/Test.sol";
+
+contract EchidnaFuzzResourceMetering is ResourceMetering, StdUtils {
+    bool failedMaxGasPerBlock;
+    bool failedRaiseBasefee;
+    bool failedLowerBasefee;
+    bool failedNeverBelowMinBasefee;
+    bool failedMaxRaiseBasefeePerBlock;
+    bool failedMaxLowerBasefeePerBlock;
+
+    constructor() {
+        initialize();
+    }
+
+    function initialize() internal initializer {
+        __ResourceMetering_init();
+    }
+
+    /**
+     * @notice Takes the necessary parameters to allow us to burn arbitrary amounts of gas to test
+     *         the underlying resource metering/gas market logic
+     */
+    function testBurn(uint256 _gasToBurn, bool _raiseBaseFee) public {
+        // Part 1: we cache the current param values and do some basic checks on them.
+        uint256 cachedPrevBaseFee = uint256(params.prevBaseFee);
+        uint256 cachedPrevBoughtGas = uint256(params.prevBoughtGas);
+        uint256 cachedPrevBlockNum = uint256(params.prevBlockNum);
+
+        // check that the last block's base fee hasn't dropped below the minimum
+        if (cachedPrevBaseFee < uint256(MINIMUM_BASE_FEE)) {
+            failedNeverBelowMinBasefee = true;
+        }
+        // check that the last block didn't consume more than the max amount of gas
+        if (cachedPrevBoughtGas > uint256(MAX_RESOURCE_LIMIT)) {
+            failedMaxGasPerBlock = true;
+        }
+
+        // Part2: we perform the gas burn
+
+        // force the gasToBurn into the correct range based on whether we intend to
+        // raise or lower the basefee after this block, respectively
+        uint256 gasToBurn;
+        if (_raiseBaseFee) {
+            gasToBurn = bound(
+                _gasToBurn,
+                uint256(TARGET_RESOURCE_LIMIT),
+                uint256(MAX_RESOURCE_LIMIT)
+            );
+        } else {
+            gasToBurn = bound(_gasToBurn, 0, uint256(TARGET_RESOURCE_LIMIT));
+        }
+
+        _burnInternal(uint64(gasToBurn));
+
+        // Part 3: we run checks and modify our invariant flags based on the updated params values
+
+        // Calculate the maximum allowed baseFee change (per block)
+        uint256 maxBasefeeChange = cachedPrevBaseFee / uint256(BASE_FEE_MAX_CHANGE_DENOMINATOR);
```

Comment:
I'm actually not sure if it's this, or per [the spec](https://github.com/ethereum-optimism/optimism/blob/dfdf9732d8d558c2c873ee267d625616f0fbc220/specs/guaranteed-gas-market.md?plain=1#L100), this: 

```suggestion
        uint256 maxBasefeeChange = cachedPrevBaseFee * (1 -  1 / uint256(BASE_FEE_MAX_CHANGE_DENOMINATOR));
```

Reply:
`cachedPrevBaseFee / uint256(BASE_FEE_MAX_CHANGE_DENOMINATOR)` is correct. Your maximum change is `1/denominator` of the current base fee.